derive/src/trait_bounds.rs

```diff
@@ -0,0 +1,160 @@
+// Copyright 2019-2020 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use alloc::vec::Vec;
+use proc_macro2::Ident;
+use syn::{
+    parse_quote,
+    spanned::Spanned,
+    visit::Visit,
+    Generics,
+    Result,
+    Type,
+};
+
+/// Visits the ast and checks if one of the given idents is found.
+struct ContainIdents<'a> {
+    result: bool,
+    idents: &'a [Ident],
+}
+
+impl<'a, 'ast> Visit<'ast> for ContainIdents<'a> {
+    fn visit_ident(&mut self, i: &'ast Ident) {
+        if self.idents.iter().any(|id| id == i) {
+            self.result = true;
+        }
+    }
+}
+
+/// Checks if the given type contains one of the given idents.
+fn type_contains_idents(ty: &Type, idents: &[Ident]) -> bool {
+    let mut visitor = ContainIdents {
+        result: false,
+        idents,
+    };
+    visitor.visit_type(ty);
+    visitor.result
+}
+
+/// Adds a `TypeInfo + 'static` bound to all relevant generic types, correctly
+/// dealing with self-referential types.
+pub fn add(input_ident: &Ident, generics: &mut Generics, data: &syn::Data) -> Result<()> {
+    let ty_params = generics.type_params_mut().fold(Vec::new(), |mut acc, p| {
+        p.bounds.push(parse_quote!(::scale_info::TypeInfo));
+        p.bounds.push(parse_quote!('static));
+        acc.push(p.ident.clone());
+        acc
+    });
+
+    if ty_params.is_empty() {
+        return Ok(())
+    }
+
+    let types = collect_types_to_bind(input_ident, data, &ty_params)?;
+
+    if !types.is_empty() {
+        let where_clause = generics.make_where_clause();
+
+        types.into_iter().for_each(|ty| {
```

Comment:
An alternative API would be to return these types and have the caller add them to the where clause, rather than passing `&mut Generics`. Not saying it's better just thinking out loud.

Reply:
Yeah it's a toss-up. Passing around a `&mut` is a bit iffy, but keeps the caller short(er). Moving this (and another snippet) to the caller would make the `trait_bounds` be less coupled to `TypeInfo` stuff which might be valuable if we ever need similar kinds of type filtering for other purposes.
I'd be inclined to keep it as-is for now, mostly because I'm lazy; if you disagree I'll move it.

Reply:
Fine by me